File: go/closure/polymorphism/mammal_test.go
package polymorphism

import (
	"fmt"
	"testing"
)

func TestCow(t *testing.T) {
	c := Cow()
	c.print()
	c.speak()

	fmt.Printf("Cow weight: %d\n", c.getWeight())
	if c.getWeight() != 100 {
		t.Errorf("Cow weight initialization is wrong.")
	}

	c.setWeight(200)
	fmt.Printf("Cow new weight: %d\n", c.getWeight())
	if c.getWeight() != 200 {
		t.Errorf("Cow setWeight is wrong.")
	}
}

func TestSheep(t *testing.T) {
	s := Sheep()
	s.print()
	s.speak()

	fmt.Printf("Sheep weight: %d\n", s.getWeight())
	if s.getWeight() != 50 {
		t.Errorf("Sheep weight initialization is wrong.")
	}

	s.setWeight(100)
	fmt.Printf("Sheep new weight: %d\n", s.getWeight())
	if s.getWeight() != 100 {
		t.Errorf("Sheep setWeight is wrong.")
	}
}

File: cpp/src/closure/polymorphism/mammal_test.cpp
#include "mammal.hpp"

void test_cow();
void test_sheep();

int main() {
    test_cow();
    test_sheep();

    return 0;
}

void test_cow() {
    Mammal *c = make_cow();
    std::cout << "Cow Weight: " << c->getWeight() << '\n';
    c->print();
    c->speak();

    std::cout << "Cow Weight: " << c->getWeight() << '\n';
    assert(c->getWeight() == 100);

    c->setWeight(200);
    std::cout << "Cow new weight: " << c->getWeight() << '\n';
    assert(c->getWeight() == 200);

    delete_cow(c);
}

void test_sheep() {
    Mammal *s = make_sheep();
    std::cout << "Sheep Weight: " << s->getWeight() << '\n';
    s->print();
    s->speak();

    std::cout << "Sheep Weight: " << s->getWeight() << '\n';
    assert(s->getWeight() == 50);

    s->setWeight(100);
    std::cout << "Sheep new weight: " << s->getWeight() << '\n';
    assert(s->getWeight() == 100);

    delete_sheep(s);
}
